mullvad-vpn: build from source, inherit mullvad version

aktaboot opened the pull request:
Thanks to @TomaSajt for  previous work

The version is now inherited from mullvad's version

I have been using this for some months now

Closes: #424929
Closes: #474877
Closes: #521059
Closes: #522471

Edited by @SigmaSquadron: This is pre-AI-policy, but I assert that no LLM assistance was used anywhere in this PR by myself or @jackrosenberg. It may be best to review this by looking at the file without the diff, as this might as well be a completely new package.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).


---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

aktaboot committed:
mullvad-vpn: build from source, inherit mullvad version

Co-authored-by: Fernando Rodrigues <alpha@sigmasquadron.net>
Co-authored-by: Jack Rosenberg <github@jackr.eu>
Co-authored-by: TomaSajt <62384384+TomaSajt@users.noreply.github.com>
Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

jackrosenberg commented on pkgs/by-name/mu/mullvad-vpn/package.nix:
```diff
@@ -1,186 +1,104 @@
-      sigmasquadron
+      Br1ght0ne
+      ymarkus
+      ataraxiasjel
```
please keep me as maintainer :)

jackrosenberg reviewed (commented):
Anything I can do to help this out?

SigmaSquadron committed:
nixos/mullvad-vpn: split daemon package from GUI package, refactor

pkgs.mullvad-vpn no longer contains the Daemon, so this split is necessary
to ensure the VPN systemd service continues working. Also refactors the module
in order to make it easier to work with.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
nixos/release-notes: add entry regarding the mullvad GUI migration

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron commented on nixos/modules/services/networking/mullvad-vpn.nix:
```diff
@@ -5,103 +5,114 @@
   };
 
   config = mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = !(cfg.package.hasMullvadDaemon or false);
+        message = "'services.mullvad-vpn.package' is using '${
+          cfg.package.name or "unknown-package"
+        }', but this package has indicated that it does not contain a Mullvad VPN Daemon. This may be due to the fact that you have configured 'services.mullvad-vpn.package' to point to 'pkgs.mullvad-vpn', which no longer contains the Mullvad Daemon. Please unset 'services.mullvad-vpn.package' and enable 'services.mullvad-vpn.gui.enable' if you wish to use Mullvad VPN's graphical user interface." ...
+      }
+    ];
+
+    warnings =
+      optional (!(cfg.gui.package.hasMullvadGUI or false))
+        "'services.mullvad-vpn.gui.package' reports that it does not contain a Mullvad VPN graphical user interface. The Mullvad VPN desktop app may be unavailable.";
+
```
Reviewers: This is a guardrail to prevent users migrating to 26.11 from accidentally breaking their Mullvad configurations if they have the main package set to `mullvad-vpn`, as it no longer contains the daemon. It works by checking some `passthru` values present in each of the Mullvad packages. `hasMullvadDaemon` is being added in #545163.

SigmaSquadron replied:
There was a bug in the assertion logic which is now fixed.

SigmaSquadron commented on nixos/modules/services/networking/mullvad-vpn.nix:
```diff
@@ -5,103 +5,114 @@
-      after = [
-        "network-online.target"
-        "NetworkManager.service"
-        "systemd-resolved.service"
-      ]
-      ++ lib.optional cfg.enableEarlyBootBlocking "mullvad-early-boot-blocking.service";
-      # See https://github.com/NixOS/nixpkgs/issues/262681
-      path = lib.optional config.networking.resolvconf.enable config.networking.resolvconf.package;
-      startLimitBurst = 5;
-      startLimitIntervalSec = 20;
-      serviceConfig = {
-        ExecStart = "${cfg.package}/bin/mullvad-daemon -v --disable-stdout-timestamps";
-        Restart = "always";
-        RestartSec = 1;
+      mullvad-daemon = {
+        description = "Mullvad VPN daemon";
+        wantedBy = [ "multi-user.target" ];
+        wants = [
+          "network.target"
+          "network-online.target"
+        ];
+        after = [
+          "network-online.target"
+          "NetworkManager.service"
+          "systemd-resolved.service"
+        ]
+        ++ optional cfg.enableEarlyBootBlocking "mullvad-early-boot-blocking.service";
+
+        path =
+          # Necessary for certain obfuscation types and DAITA.
+          singleton pkgs.iproute2
+          # See https://github.com/NixOS/nixpkgs/issues/262681
+          ++ optional config.networking.resolvconf.enable config.networking.resolvconf.package;
+        startLimitBurst = 5;
+        startLimitIntervalSec = 20;
+        serviceConfig = {
+          ExecStart = "${getExe' cfg.package "mullvad-daemon"} -v --disable-stdout-timestamps";
+          Restart = "always";
+          RestartSec = 1;
+        };
```
Reviewers: The large diff is due to my decision to remove the repeated `systemd.service` key. Here are the actual modifications:
- Swapped hardcoded binary paths with `getExe'`
- Added `iproute2` to `$PATH`, as required by DAITA and some obfuscation methods.

jackrosenberg replied:
iproute2 will be phased out for obfuscation in https://github.com/mullvad/mullvadvpn-app/pull/10794, but is still required for DAITA

SigmaSquadron replied:
Wonderful. We can drop iproute2 with whatever update removes it upstream. I'll add a comment.

jackrosenberg replied:
Does DAITA work currently for us?

SigmaSquadron replied:
I've confirmed DAITA to work correctly on 2026.3 with these changes.

SigmaSquadron commented on nixos/modules/services/networking/mullvad-vpn.nix:
```diff
@@ -5,103 +5,114 @@
+    enableExcludeWrapper =
+      (mkEnableOption "the security wrapper the allows the use of `mullvad-exclude`, a command that tunnels a single command directly to the clearnet, bypassing the system-wide VPN. It is recommended to disable this option for security models where privileged (`setuid`) binaries are a concern")
+      // {
+        default = true;
+      };
 
-    enableEarlyBootBlocking = mkOption {
-      type = types.bool;
-      default = false;
-      description = ''
-        This option activates an additional oneshot systemd service to ensure that the mullvad daemon
-        will start and block traffic before any network configuration will be applied.
-        This matches what upstream Mullvad distributes for their supported distros, but is disabled by
-        default in NixOS as it may conflict with non-Mullvad network configuration.
-      '';
-    };
+    enableEarlyBootBlocking = mkEnableOption "an additional systemd service that ensures that the Mullvad VPN daemon will block all traffic early in the boot process, as opposed to blocking traffic after the network has been set up. Enabling this option matches what upstream Mullvad distributes for their supported distros, but is disabled by default in NixOS as it may conflict with non-Mullvad ne ...
 
-    package = mkPackageOption pkgs "mullvad" {
-      example = "pkgs.mullvad-vpn";
-      extraDescription = ''
-        `pkgs.mullvad` only provides the CLI tool, `pkgs.mullvad-vpn` provides both the CLI and the GUI.
-      '';
+    gui = {
+      enable = mkEnableOption "the Mullvad VPN graphical user interface";
+      package = mkPackageOption pkgs "mullvad-vpn" { };
```
Reviewers: Swapped the order of the `package` option in the module file, and updated some descriptions so they fit better with `mkEnableOption`. We no longer use custom `mkOption`s.

SigmaSquadron committed:
mullvad: run all tests and add meta.platforms attribute

These tests now work on the latest version, so we can stop skipping them.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

jackrosenberg commented on nixos/modules/services/networking/mullvad-vpn.nix:
```diff
@@ -5,103 +5,115 @@
   config = mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = cfg.package.hasMullvadDaemon or false;
```
assertion triggers properly on rebuild with old config.

jackrosenberg reviewed (approved):
Tested on laptop building off of this branch, DAITA + multihop + obfuscation works. Nice job people :)

jackrosenberg commented on nixos/doc/manual/release-notes/rl-2611.section.md:
```diff
@@ -75,6 +75,8 @@
 - Rustical migrates from `settings.http.host` and `settings.http.port` to `settings.http.bind` to support UNIX domain sockets as well as TCP sockets in one setting.
 
+- The Mullvad VPN service now has a separate toggle to enable the Mullvad VPN graphical user interface. If you have previously used Mullvad on a desktop by setting `services.mullvad-vpn.package` to `pkgs.mullvad-vpn`, you should now **unset that option**, and enable `services.mullvad-vpn.gui.enable`. The VPN will not work if `services.mullvad-vpn.package` is set to `pkgs.mullvad-vpn`, as `pkgs.mu ...
+
```
we'll need to change the wiki to reflect this once it goes thru

jackrosenberg commented on nixos/modules/services/networking/mullvad-vpn.nix:
```diff
@@ -5,103 +5,115 @@
+        after = [
+          "network-online.target"
+          "NetworkManager.service"
+          "systemd-resolved.service"
```
systemd.resolved is false by default ([1](https://search.nixos.org/options?channel=unstable&query=resolved.enable&type=options)). 

```suggestion
        ++ optional config.services.resolved.enable 
          "systemd-resolved.service"
```

SigmaSquadron replied:
Mullvad depends on resolved if I'm not mistaken so this should probably not be an optional, but the module should just enable resolved

jackrosenberg replied:
Ah. I'm more in favor of an assertion then. These things should be opt in explicitly IMO.

SigmaSquadron replied:
I did that and then I realised an assertion wouldn't be correct here as Mullvad technically works without resolved, it just prefers to use it, thus `lib.mkDefault` is ideal.
